### ISSUE TITLE
QGIS crashes while calling readLayerXml() on a WMS layer on Windows

### ISSUE THREAD

**Issue by cxcandid** [qgis/QGIS#46741]
### What is the bug or the crash?

QGIS crashes while calling readLayerXml() on Windows.

### Steps to reproduce the issue

Try the following code, which is part of the *changeDataSource* plugin, to reproduce the bug:

rasterLayer ... active WMS raster layer

```python
from PyQt5.QtXml import *

rasterLayer = iface.activeLayer()

XMLDocument = QDomDocument("style")
XMLMapLayers = XMLDocument.createElement("maplayers")
XMLMapLayer = XMLDocument.createElement("maplayer")
context = QgsReadWriteContext()
rasterLayer.writeLayerXml(XMLMapLayer,XMLDocument, context)

#   XMLMapLayer.firstChildElement("datasource").firstChild().setNodeValue(newDatasource)
#   XMLMapLayer.firstChildElement("provider").firstChild().setNodeValue(newProvider)

XMLMapLayers.appendChild(XMLMapLayer)
XMLDocument.appendChild(XMLMapLayers)
rasterLayer.readLayerXml(XMLMapLayer, context) # <== QGIS crashes here
```

### Versions

QGIS version
3.22.2-Białowieża
QGIS code revision
1601ec46d0
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.4.0
PROJ version
8.2.0
EPSG Registry database version
v10.038 (2021-10-21)
GEOS version
3.10.0-CAPI-1.16.0
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 2009

Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.5
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context
Crash ID: cabb170ce9b4cbfc835aad0515873f11dc7a8376

Stack Trace

QgsDefaultRasterLayerLegend::createLayerTreeModelLegendNodes :
QgsLayerTreeModel::addLegendToLayer :
QgsLayerTreeModel::refreshLayerLegend :
QgsLayerTreeModel::layerLegendChanged :
QObject::qt_static_metacall :
QObject::qt_static_metacall :
QObject::qt_static_metacall :
QgsRasterLayer::setRendererForDrawingStyle :
QgsRasterLayer::setDataProvider :
QgsRasterLayer::readXml :
QgsMapLayer::readLayerXml :
PyInit__core :
PyArg_ParseTuple_SizeT :
PyEval_EvalFrameDefault :
PyObject_GC_Del :
PyEval_EvalCodeWithName :
PyEval_EvalCodeEx :
PyEval_EvalCode :
PyEval_GetBuiltins :
PyEval_GetBuiltins :
PyEval_EvalFrameDefault :
PyFloat_FromDouble :
PyEval_EvalFrameDefault :
PyObject_GC_Del :
PyFloat_FromDouble :
PyEval_EvalFrameDefault :
PyObject_GC_Del :
PyFloat_FromDouble :
PyEval_EvalFrameDefault :
PyFloat_FromDouble :
PyEval_EvalFrameDefault :
PyFloat_FromDouble :
PyEval_EvalFrameDefault :
PyFunction_Vectorcall :
PyFloat_FromDouble :
PyVectorcall_Call :
PyObject_Call :
PyInit__gui :
QWidget::event :
QFrame::event :
QAbstractScrollArea::event :
PyInit__gui :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processKeyEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :

QGIS Info
QGIS Version: 3.22.2-Bia?owie?a
QGIS code revision: 1601ec46d0
Compiled against Qt: 5.15.2
Running against Qt: 5.15.2
Compiled against GDAL: 3.4.0
Running against GDAL: 3.4.0

System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.19043

**Comment by roya0045** [comment]
The title and descriptions mentions the crash on write but the code snippets mentions a crash on read. Also is there a stacktrace or anything more?

**Comment by cxcandid** [comment]
@roya0045 this was a mistake, should be *readLayerXml()* of course.
BTW: the code is taken from *changeDataSource* plugin, which fails in QGIS 3.22.2 as well.
In QGIS 3.20.3 everything is fine. I can add a stack trace, but I guess it won't be of much help.

**Comment by cxcandid** [comment]
@roya0045 in case it might be helpful after all, I added the stack trace to the original report.

**Comment by cxcandid** [comment]
QGIS still crashes in 3.22.3 while trying to change a raster datasource with *changeDataSource* plugin.

**Comment by elpaso** [comment]
Cannot reproduce on linux/master.

![immagine](https://user-images.githubusercontent.com/142164/150765795-a9d5b905-12a8-4b07-8944-2e2c6c203a86.png)

**Comment by cxcandid** [comment]
I installed QGIS 3.22.3 on a brand new Lenovo Notebook with Windows 10 Home (20H2) and received the same crash.
Can someone please confirm this QGIS 3.22.x bug on Windows?

**Comment by cxcandid** [comment]
This issue still exists in QGIS 3.24.0 (Windows 10).
QGIS throws "Windows fatal exception: access violation".

**Comment by gioman** [comment]
> This issue still exists in QGIS 3.24.0 (Windows 10).

@cxcandid no crash here (master, Win10) with the code you provided.

**Comment by cxcandid** [comment]
QGIS crashes if source of raster layer is WMS. With file based images it's ok.

**Comment by gioman** [comment]
> QGIS crashes if raster layer source is WMS or XYZ

Confirmed.

**Comment by cxcandid** [comment]
Issue is solved in QGIS 3.24.1 Rev. 5709b824. 

Thanks a lot!
